### PR TITLE
Add support for naming experiments and jobs 

### PR DESCRIPTION
**Pull request by LizaShak**
This PR is allowing the researcher to set up a name for experiment and the job run for research ease of tracking and clarity.

Technical Notes:
-  Config.json includes new fields: experiment_name, job_name, job_description.
-  Rename prefix_name to index_prefix_name in config.json for clarity
-  Backwards compatible - if fields are not set, use current logic
-  Job runs will add dates and times to the name

## pros
- Enables to group all relevant job run under the same experiment name
- Provide reasonable name for an experiment
- Decouple experiment from index name
- Allow easy export and comparison for the whole experiment - look below an example
- Backward compatible

## cons
- Unaware of index which the job ran against - will be fixed by adding index_name tag - #480 

Example for export as Markdown (related to #319)

### Metrics outputs:

Metric|baseline|H1 - generated_answer|H2 - generated_hypothetical_document_to_answer|H4 - generated_related_questions_azure_search_reranking_filter_questions|H3 - generated_related_questions_azure_search_reranking
---|---|---|---|---|---
Context relevancy|21.101|21.137 (0.04) ↑|21.222 (0.12) ↑|21.094 (-0.01) ↓|21.232 (0.13) ↑
Answer correctness|56.215|53.83 (-2.39) ↓|53.535 (-2.68) ↓|45.516 (-10.70) ↓|56.171 (-0.04) ↓
Faithfulness|69.062|66.531 (-2.53) ↓|64.987 (-4.08) ↓|48.544 (-20.52) ↓|68.636 (-0.43) ↓
Context recall|96.619|96.451 (-0.17) ↓|96.752 (0.13) ↑|96.788 (0.17) ↑|96.792 (0.17) ↑
Answer relevancy|90.607|89.877 (-0.73) ↓|89.937 (-0.67) ↓|73.219 (-17.39) ↓|90.442 (-0.17) ↓
Context precision|59.895|60.357 (0.46) ↑|60.275 (0.38) ↑|60.687 (0.79) ↑|60.786 (0.89) ↑
Answer similarity|92.049|91.703 (-0.35) ↓|91.804 (-0.25) ↓|89.388 (-2.66) ↓|92.228 (0.18) ↑
Description|baseline|generate an hypothetical answer the question for document retrieval|generated an hypothetical document to answer the question for document retrieval|generated related questions based on user query for document retrieval, with non related question filtering plus reranking using azure search built in scoring algorithm|generated related questions based on user query for document retrieval plus reranking using azure search built in scoring algorithm

This PR also fixes spacy dependency issue which causes test to run much longer and fail in timeout by downloading it only once. and disable sampling in index test, as it takes too long to run.

### CODE REVIEW

**Comment by guybartal** [comment]
@LizaShak, in the PR description you wrote `Job runs will add dates and times to the name` but I couldn't see anything related to that in the code. Are you sure you handled it?
this is important in case someone reruns a job and wouldn't want the previous run to get overwritten.

**Comment by tanya-borisova** [comment]
I like the idea of this, but for me it isn't immediately clear what is "experiment name" and what is "job name", maybe we could make the names a little clearer, such as "mlflow_expeirment_name" and "azureml_job_name", if that is what they are?

**Comment by LizaShak** [comment]
> azureml_job_name



> I like the idea of this, but for me it isn't immediately clear what is "experiment name" and what is "job name", maybe we could make the names a little clearer, such as "mlflow_expeirment_name" and "azureml_job_name", if that is what they are?

Hi @tanya-borisova , I am not sure about this, because we only support azureml, and mlflow is used to create and experiment in azureml... So it just quite obvious to me, if you think we still need to change it maybe just prefix both with azureml?

**Comment by guybartal** [comment]
I think this requires more work, for instance, if we set several config.json files and/or one/more of them use multiple values for chunking strategies like `overlap_size` or `overlap_size` then this would yield several job runs under the same name and that would be weird.
at least lets make sure all of those configurable hyper-parameters are logged in the job run using ml-flow.
and we should consider adding some sort of smart suffix for those job runs.

File: rag_experiment_accelerator/config/tests/data/config.json [hunk 1]
```diff
@@ -1,4 +1,10 @@
 {
+<<<<<<< HEAD
```
Comment (guybartal) [inline rag_experiment_accelerator/config/tests/data/config.json:2]: merge leftovers? 😮

File: rag_experiment_accelerator/config/config.py [hunk 2]
```diff
@@ -74,7 +77,10 @@ def __init__(
         self.OVERLAP_SIZES = config_json["chunking"]["overlap_size"]
         self.EF_CONSTRUCTIONS = config_json["ef_construction"]
         self.EF_SEARCHES = config_json["ef_search"]
-        self.NAME_PREFIX = config_json["name_prefix"]
+        self.INDEX_NAME_PREFIX = config_json["index_name_prefix"]
+        self.EXPERIMENT_NAME = config_json["experiment_name"] or self.INDEX_NAME_PREFIX
+        self.JOB_NAME = config_json["job_name"]
```
Comment (guybartal) [inline rag_experiment_accelerator/config/config.py:82]: I think either add suffix for each job name according to the current `index_config` combination, or simply add incremental number.
more other all mlflow tags which are not computational show move to the first step from the last step (eval).

**Comment by guybartal** [comment]
pros and cons added to the PR description as requested in the review meeting.

**Review by guybartal — APPROVED** [review]

**Pull request merged**